add delete dataflow in architecture document (#591)
vdaas/vald: docs/overview/architecture.md

<!-- ### Update -->

<!-- ### Upsert -->

### Delete

<img src="../../assets/docs/delete_flow.png" />

When the user delete a vector which is indexed in Vald Agent:

1. Vald Ingress receives the delete request from the user. The request includes the vector ID(s) which is specified by the user.
2. Vald Ingress will forward the request to the Vald Filter Gateway.

> 3. Vald Filter Gateway will forward the request to the Vald Meta Gateway. Vald Meta Gateway is used to resolve the internal used UUID from Vald Meta to the user inserted vector ID in the Insert Step.

Reviewer: ```suggestion
3. Vald Filter Gateway will forward the request to the Vald Meta Gateway. Vald Meta Gateway is used to resolve the internal used UUID of the vector from Vald Meta.
```

> 4. Vald Meta Gateway will forward the request to the Vald Meta to confirm whether the metadata, which contains the request ID(s), is exist or not.

Reviewer: ```suggestion
4. Vald Meta Gateway will forward the request to the Vald Meta to confirm whether the metadata, which contains the request ID(s), exist or not.
```

> 4. Vald Meta Gateway will forward the request to the Vald Meta to confirm whether metadata which contains request ID(s) is existing or not.

Reviewer: ```suggestion
4. Vald Meta Gateway will forward the request to the Vald Meta to confirm whether the metadata, which contains the request ID(s), is exist or not.
```

> 5. Vald Meta get the UUID(s) by the request ID(s). Return error if no UUID(s) is found, return with the error.

Reviewer: ```suggestion
5. Vald Meta gets the UUID(s) by the request ID(s). Return error if no UUID(s) is found, return with the error.
```

> 5. Vald Meta get the UUID(s) from the request ID(s). When the UUID(s) be not found, return with the error.

Reviewer: ```suggestion
5. Vald Meta get the UUID(s) by the request ID(s). Return error if no UUID(s) is found.
```

> 6. If Vald Meta Gateway gets the UUID(s), Vald Meta Gateway will forward to the request with the UUID(s) to the Vald LB Gateway through the Vald Backup Gateway.

Reviewer 2: I think number 6 is not a connection to LB Gateway, but a connection to Backup Gateway:thinking:

Reviewer 2: Confirm Location information with Backup Gateway and send Remove request to LB Gateway. Then, when the request is successful, the Backup data is deleted.

Author: Now, you're right that Vald Backup Gateway will get Agent IP using UUID.
https://github.com/vdaas/vald/blob/master/pkg/gateway/vald/handler/grpc/handler.go#L741

But, in this article, we're targeting the new architecture, like #579 .(I cannot source code...)
And, the author, @kpango told us that getting location info from Vald Backup Gateway is too cost so he decide broadcast all agent with UUID instead of that.

> 7. Vald LB Gateway will broadcast the request with UUID(s) to the Vald Agents. Each Vald Agent will delete the vector data and the medatada if the corresponding UUID(s) is found in the onmemory graph index.

Reviewer: ```suggestion
7. Vald LB Gateway will broadcast the request with UUID(s) to the Vald Agents. Each Vald Agent will delete the vector data and the metadata if the corresponding UUID(s) is found in the on memory graph index.
```

Author: @kevindiu  I think `on memory` is not correct grammar 🤔

Author: ```suggestion
7. Vald LB Gateway will broadcast the request with UUID(s) to the Vald Agents. Each Vald Agent will delete the vector data and the metadata if the corresponding UUID(s) is found in the on-memory graph index.
```
how about it?

Reviewer: may be `in memory`?

Reviewer: @vankichi `in-memory` should be correct.
https://en.wikipedia.org/wiki/In-memory_database

Author: maybe typographical issues 😅 
I'm not sure what is correct...

Reviewer: I googled  `on memory graph index`  and seems no one use `on memory`. 
most of the result use `in-memory`

Author: ```suggestion
7. Vald LB Gateway will broadcast the request with UUID(s) to the Vald Agents. Each Vald Agent will delete the vector data and the metadata if the corresponding UUID(s) is found in the in-memory graph index.
```
Good?

Reviewer: I think it is good

> 7. Vald LB Gateway will broadcast the request to the Vald Agent. All pods of Vald Agent check the vector whose UUID is requested one is indexed or not. When idexing data is existing, Vald Agent will delete the vector and the UUID in an onmemory graph index.

Reviewer: ```suggestion
7. Vald LB Gateway will broadcast the request with the UUID to Vald Agents. Each Vald Agent will delete the vector data and the metadata if the corresponding UUID is found in the on-memory graph index.
```

> 8. If Vald Agent successfully deletes the request data, it will return success to the Vald LB Gateway.

Reviewer 2: ```suggestion
8. If Vald Agent successfully deletes the request data, it will return ip addresses and name of Vald Agent to the Vald LB Gateway.
```

9. After Vald LB Gateway receives success from the Vald Agents, Vald LB Gateway will return success to the Vald Backup Gateway.
10. Vald Backup Gateway will forward the request with the UUID to the Vald Compressor.
11. Vald Compressor will forward the UUID(s) to the Vald Backup Manager.

> 12. Vald Backup Manager will delete the data whose UUID is the requested UUID(s).

Reviewer: ```suggestion
12. Vald Backup Manager will delete the data with the same UUID.
```

> 13. If deleting backup related to the request is an success, Vald Backup Gateway will return success to the Vald Meta Gateway.

Reviewer: ```suggestion
13. Vald Backup Gateway will return success to the Vald Meta Gateway if the data is successfully deleted from the Vald Backup Manager.
```

14. Vald Meta Gateway will forward the UUID(s) to the Vald Meta.

> 15. Vald Meta will delete the data from the persistent layer if hte data is successfully deleted from the Vald Agents, and store the deleted UUID(s) and the deletion time in the persistent layer.

Reviewer: ```suggestion
15. Vald Meta will delete the data from the persistent layer if the data is successfully deleted from the Vald Agents, and store the deleted UUID(s) and the deletion time in the persistent layer.
```

> 15. Vald Meta will delete the UUID(s) and vector ID(s) that were successfully processed by the Vald Agent(s) to the persistent layer such as Redis, Cassandra, MySQL, etc. And, after successfully to delete, Vald Meta also will store delete UUID and Time.

Reviewer: ```suggestion
15. Vald Meta will delete the data from the persistent layer if the data is successfully deleted from the Vald Agents, and store the deleted UUID and the deletion time in the persistent layer.
```

<!-- ### Update -->

<!-- ### Upsert -->

### Delete

<img src="../../assets/docs/delete_flow.png" />

When the user deletes a vector which is indexed in Vald Agent:

> 1. Vald Ingress receives the delete request from the user. The request includes the vector ID(s) which is specified by the user.

Reviewer 3: ```suggestion
1. Vald Ingress receives the delete request from the user. The request includes the vector ID(s), which is specified by the user.
```

2. Vald Ingress will forward the request to the Vald Filter Gateway.
3. Vald Filter Gateway will forward the request to the Vald Meta Gateway. Vald Meta Gateway is used to resolve the internal used UUID of the vector from the Vald Meta.

> 4. Vald Meta Gateway will forward the request to the Vald Meta to confirm whether the metadata, which contains the request ID(s), exist or not.

Reviewer 3: ```suggestion
4. Vald Meta Gateway will forward the request to the Vald Meta to confirm whether the metadata, which contains the request ID(s), exists or not.
```

5. Vald Meta gets the UUID(s) by the request ID(s). Return error if no UUID(s) is found, return with the error.

> 6. If Vald Meta Gateway gets the UUID(s), Vald Meta Gateway will forward to the request with the UUID(s) to the Vald Backup Gateway.

Reviewer 3: ```suggestion
6. If Vald Meta Gateway gets the UUID(s), Vald Meta Gateway will forward it to the request with the UUID(s) to the Vald Backup Gateway.
```

7. Vale Backup Gateway will forward the request with the UUID(s) to the Vald LB Gateway.
8. Vald LB Gateway will broadcast the request with UUID(s) to the Vald Agents. Each Vald Agent will delete the vector data and the metadata if the corresponding UUID(s) is found in the in-memory graph index.
9. If Vald Agent successfully deletes the request data, it will return success to the Vald LB Gateway.

> 10. After Vald LB Gateway receives success with location info (e.g. IP address of pod) from the Vald Agent, Vald LB Gateway will return success to the Vald Backup Gateway.

Reviewer 3: ```suggestion
10. After Vald LB Gateway receives success with location info (e.g, IP address of pod) from the Vald Agent, Vald LB Gateway will return success to the Vald Backup Gateway.
```

11. Vald Backup Gateway will forward the request with the UUID to the Vald Compressor.
12. Vald Compressor will forward the UUID(s) to the Vald Backup Manager.
13. Vald Backup Manager will delete the data with the same UUID(s).
14. Vald Backup Gateway will return success to the Vald Meta Gateway if the data is successfully deleted from the Vald Backup Manager.
15. Vald Meta Gateway will forward the UUID(s) to the Vald Meta.

> 16. Vald Meta will delete the data from the persistent layer if the data is successfully deleted from the Vald Agents, and store the deleted UUID(s) and the deletion time in the persistent layer.

Reviewer 3: ```suggestion
16. Vald Meta will delete the data from the persistent layer if it is successfully deleted from the Vald Agents, store the deleted UUID(s) and the deletion time in the persistent layer.
```
